src/vmem.cc

```diff
@@ -86,10 +113,52 @@ uint64_t VirtualMemory::get_offset(champsim::address vaddr, std::size_t level) c
 
 uint64_t VirtualMemory::get_offset(champsim::page_number vaddr, std::size_t level) const { return get_offset(champsim::address{vaddr}, level); }
 
-champsim::page_number VirtualMemory::ppage_front() const
+// champsim::page_number VirtualMemory::ppage_front() const
+// {
+//   assert(available_ppages() > 0);
+//   return ppage_free_list.front();
+// }
+
+
+std::deque<std::pair<champsim::page_number, bool>>::iterator
+VirtualMemory::ppage_index(uint32_t cpu_num, champsim::page_number vaddr)
 {
-  assert(available_ppages() > 0);
-  return ppage_free_list.front();
+    auto vaddr_val = vaddr.to<uint64_t>();
+    std::size_t idx = (hash_constants[cpu_num] ^ vaddr_val) % ppage_free_list.size();
+
+    auto it = ppage_free_list.begin() + idx;
+    
+    // for keeping track of the original ptr
+    auto start_itr = it;
+    while(it != ppage_free_list.end() && it->second){
+      it++;
+    }
```

Comment:
```suggestion
    auto start_itr = ppage_free_list.begin() + idx;
    
    auto it = std::find_if(start_itr, ppage_free_list.end(), [](const auto& free_list_entry) { return !free_list_entry.second; });
```

There's a standard library function for what you're doing here.